parquet-hadoop/src/main/java/org/apache/parquet/crypto/AesEncryptor.java

```diff
@@ -0,0 +1,232 @@
+    }
+    else {
+      tagLength = 0;
+      try {
+        aesCipher = Cipher.getInstance("AES/CTR/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Failed to create CTR cipher", e);
+      }
+      ctrIV = new byte[CTR_IV_LENGTH];
+      Arrays.fill(ctrIV, (byte) 0);
+      // Setting last bit of initial CTR counter to 1
+      ctrIV[CTR_IV_LENGTH - 1] = (byte) 1;
+    }
+    
+    localNonce = new byte[NONCE_LENGTH];
+  }
+
+  @Override
+  public byte[] encrypt(byte[] plainText, byte[] AAD)  throws IOException {
+    return encrypt(true, plainText, AAD);
+  }
+  
+  /**
+   * 
+   * @param writeLength whether to write buffer length (4-byte little endian)
+   * @param plainText
+   * @param AAD
+   * @return
+   * @throws IOException
+   */
+  public byte[] encrypt(boolean writeLength, byte[] plainText, byte[] AAD)  throws IOException {
+    randomGenerator.nextBytes(localNonce);
+    return encrypt(writeLength, plainText, localNonce, AAD);
+  }
+  
+  /**
+   * 
+   * @param writeLength
+   * @param plainText
+   * @param nonce can be generated locally or supplied. the latter is needed for footer signature verification.
+   * @param AAD
+   * @return
+   * @throws IOException
+   */
+  public byte[] encrypt(boolean writeLength, byte[] plainText, byte[] nonce, byte[] AAD)  throws IOException {
+    if (nonce.length != NONCE_LENGTH) throw new IOException("Wrong nonce length " + nonce.length);
+    int plainTextLength = plainText.length;
+    int cipherTextLength = NONCE_LENGTH + plainTextLength + tagLength;
+    int lengthBufferLength = (writeLength? INT_LENGTH: 0);
+    byte[] cipherText = new byte[lengthBufferLength + cipherTextLength];
+    int inputLength = plainTextLength;
+    int inputOffset = 0;
+    int outputOffset = lengthBufferLength + NONCE_LENGTH;
+    try {
+      if (Mode.GCM == aesMode) {
+        GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH * 8, nonce);
+        aesCipher.init(Cipher.ENCRYPT_MODE, aesKey, spec);
+        if (null != AAD) aesCipher.updateAAD(AAD);
+      }
+      else {
+        System.arraycopy(nonce, 0, ctrIV, 0, NONCE_LENGTH);
+        IvParameterSpec spec = new IvParameterSpec(ctrIV);
+        aesCipher.init(Cipher.ENCRYPT_MODE, aesKey, spec);
+      }
+      // Breaking encryption into multiple updates, to trigger h/w acceleration in Java 9-11
+      while (inputLength > CHUNK_LENGTH) {
+        int written = aesCipher.update(plainText, inputOffset, CHUNK_LENGTH, cipherText, outputOffset);
+        inputOffset += CHUNK_LENGTH;
+        outputOffset += written;
+        inputLength -= CHUNK_LENGTH;
+      }
+      aesCipher.doFinal(plainText, inputOffset, inputLength, cipherText, outputOffset);
+    }
+    catch (GeneralSecurityException e) {
+      throw new IOException("Failed to encrypt", e);
+    }
+    // Add ciphertext length
+    if (writeLength) System.arraycopy(BytesUtils.intToBytes(cipherTextLength), 0, cipherText, 0, lengthBufferLength);
```

Comment:
If 'writeLenght' is false, should 'ciphertextLength' in line 164 of AesDecryptor.java still be gotten from the cipherText?

Reply:
no. the decrypt(InputStream) is called only for encrypted Thrift structures. For them, 'writeLength' is always true.